4.x A config setting backed by the service registry can trigger native image failures during service registry start-up

## Environment Details
* Helidon Version: 4.x
* Helidon SE or Helidon MP SE
* JDK version:
* OS:

----------

## Problem Description
Declaring a config setting that is backed by the service registry can cause native image start-up failures in the service registry.
@tomas-langer suggested the workaround of manually adding a `META-INF/helidon/reflection-config.json` file, which avoids the problem but ideally no manual steps would be necessary.

In `helidon-common-concurrency-limits` PR #10464 adds this config setting to an existing config blueprint:

```java
    @Option.Singular
    @Option.RegistryService
    @Option.Configured
    @Option.Provider(value = LimitAlgorithmListenerProvider.class)
    List<LimitAlgorithmListener> listeners();
```
and adds `@Prototype.RegistrySupport` to the declaring blueprint interface. The PR also adds the new `LimitAlgorithmListenerProvider` SPI referenced in the new config setting's `@Option.Provider` annotation.

This component is pulled in by the webserver component.

The native image test executions failed with exceptions like this:
```list
./target/helidon-tests-integration-packaging-inject 

2025.08.05 10:58:27 INFO io.helidon.logging.jul.JulProvider: Logging at runtime configured using classpath: /logging.properties
Exception in thread "main" io.helidon.service.registry.ServiceRegistryException: Resolution of type "io.helidon.common.concurrency.limits.spi.LimitAlgorithmListenerProvider" to class failed.
	at io.helidon.service.registry.CoreServiceDiscovery.toClass(CoreServiceDiscovery.java:107)
	at io.helidon.service.registry.CoreServiceDiscovery$DescriptorHandlerImpl.parseServiceProvider(CoreServiceDiscovery.java:250)
	at java.base@21.0.7/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
	at java.base@21.0.7/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base@21.0.7/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base@21.0.7/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base@21.0.7/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at java.base@21.0.7/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:276)
	at java.base@21.0.7/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base@21.0.7/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base@21.0.7/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base@21.0.7/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base@21.0.7/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base@21.0.7/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base@21.0.7/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@21.0.7/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.helidon.service.registry.CoreServiceDiscovery.<init>(CoreServiceDiscovery.java:76)
	at io.helidon.service.registry.CoreServiceDiscovery.create(CoreServiceDiscovery.java:83)
	at io.helidon.service.registry.ServiceRegistryManager.create(ServiceRegistryManager.java:197)
	at io.helidon.service.registry.GlobalServiceRegistry.registry(GlobalServiceRegistry.java:65)
	at io.helidon.service.registry.Services.get(Services.java:203)
	at io.helidon.common.config.GlobalConfig.config(GlobalConfig.java:121)
	at io.helidon.tests.integration.packaging.inject.Main.main(Main.java:44)
	at java.base@21.0.7/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
	Suppressed: java.lang.ClassNotFoundException: io.helidon.common.concurrency.limits.spi.LimitAlgorithmListenerProvider
		at java.base@21.0.7/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:52)
		at java.base@21.0.7/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
		at java.base@21.0.7/java.lang.ClassLoader.loadClass(ClassLoader.java:121)
		at io.helidon.service.registry.CoreServiceDiscovery.toClass(CoreServiceDiscovery.java:101)
		... 23 more
Caused by: java.lang.ClassNotFoundException: io.helidon.common.concurrency.limits.spi.LimitAlgorithmListenerProvider
	at java.base@21.0.7/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:52)
	at java.base@21.0.7/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base@21.0.7/java.lang.ClassLoader.loadClass(ClassLoader.java:121)
	at io.helidon.service.registry.CoreServiceDiscovery.toClass(CoreServiceDiscovery.java:105)
	... 23 more

```

As Tomas suggested, adding a `src/main/resources/META-INF/helidon/native-image/reflection-config.json` file to the component allowed the native image to work:
```json
{
    "annotated":[
    ],
    "class-hierarchy": [
    ],
    "classes": [
        "io.helidon.common.concurrency.limits.spi.LimitAlgorithmListenerProvider"
    ],
    "exclude": [
    ]
}
```

Ideally this manual step should not be needed.

## Steps to reproduce
1. In the `helidon-common-concurrency-limits` component (in the PR or in `main` if the PR has been merged) remove the `reflection-config.json` file.
2. Rebuild the component.
3. `cd <top-level-helidon-dir>/tests/integration/packaging/inject`
   `mvn clean verify -Pnative-image`
4. The test fails because the created process ends with a status other than 0. 
5. `./target/helidon-tests-integration-packaging-inject` to see the stack trace pasted above.

--- a/integrations/graal/native-image-extension/src/main/java/io/helidon/integrations/graal/nativeimage/extension/HelidonReflectionFeature.java
+++ b/integrations/graal/native-image-extension/src/main/java/io/helidon/integrations/graal/nativeimage/extension/HelidonReflectionFeature.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, 2024 Oracle and/or its affiliates.
+ * Copyright (c) 2019, 2025 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -23,6 +23,7 @@ import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
@@ -30,6 +31,7 @@ import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import io.helidon.common.Reflected;
 import io.helidon.common.features.HelidonFeatures;
@@ -41,11 +43,14 @@ import io.helidon.service.registry.ServiceLoader__ServiceDescriptor;
 import io.helidon.service.registry.ServiceRegistryConfig;
 
 import io.github.classgraph.ClassGraph;
+import io.github.classgraph.Resource;
 import io.github.classgraph.ScanResult;
 import org.graalvm.nativeimage.hosted.Feature;
 import org.graalvm.nativeimage.hosted.RuntimeClassInitialization;
 import org.graalvm.nativeimage.hosted.RuntimeReflection;
 import org.graalvm.nativeimage.hosted.RuntimeResourceAccess;
+
+import static java.util.function.Predicate.not;
 
 /**
  * Feature to add reflection configuration to the image for Helidon, CDI and Jersey.
@@ -117,6 +122,8 @@ public class HelidonReflectionFeature implements Feature {
 
         // Service descriptors
         processServiceDescriptors(context);
+        // types from META-INF/helidon/service.loader
+        processServiceLoaderForRegistry(context);
 
         // JPA Entity registration
         processEntity(context);
@@ -180,6 +187,38 @@ public class HelidonReflectionFeature implements Feature {
         Set<Class<?>> subclasses = util.findSubclasses(aClass.getName());
 
         processClasses(context, subclasses);
+    }
+
+    private void processServiceLoaderForRegistry(BeforeAnalysisContext context) {
+        context.scan()
+                .getResourcesWithPath(ServiceDiscovery.SERVICES_LOADER_RESOURCE)
+                .stream()
+                .flatMap(this::parseServiceLoaderFile)
+                .flatMap(it -> findClass(context, it))
+                .forEach(context::register);
+    }
+
+    private Stream<Class<?>> findClass(BeforeAnalysisContext context, String className) {
+        Class<?> type = context.access().findClassByName(className);
+        if (type == null) {
+            // we may not have the class on the classpath... strange, but let us not fail
+            return Stream.empty();
+        }
+        return Stream.of(type);
+    }
+
+    private Stream<String> parseServiceLoaderFile(Resource serviceLoader) {
+        try {
+            String allLines = new String(serviceLoader.load(), StandardCharsets.UTF_8);
+            // similar to what we do in CoreServiceDiscovery
+            return Arrays.stream(allLines.split("\n"))
+                    .map(String::trim)
+                    .filter(not(it -> it.startsWith("#")))
+                    .filter(not(String::isEmpty));
+        } catch (IOException e) {
+            tracer.parsing(() -> "Failed to process resource " + serviceLoader, e);
+            return Stream.of();
+        }
     }
 
     private void processServiceDescriptors(BeforeAnalysisContext context) {

